### PR TITLE
Kubernetes request routing and cluster registration

### PR DESCRIPTION
**Pull request by awly**
This change has several parts: cluster registration, cache updates,
routing and a new tctl flag.

# cluster registration

Cluster registration means adding `KubernetesClusters` to `ServerSpec`
for servers with `KindKubeService`.

`kubernetes_service` instances will parse their kubeconfig or local
`kube_cluster_name` and add them to their `ServerSpec` sent to the auth
server. They are effectively declaring that "I can serve k8s requests
for k8s cluster X".

# cache updates

This is just cache plumbing for `kubernetes_service` presence, so that
other teleport processes can fetch all of kube services. It was missed
in the previous PR implementing CRUD for `kubernetes_service`.

# routing

Now the fun part - routing logic. This logic lives in
`/lib/kube/proxy/forwarder.go` and is shared by both `proxy_service`
(with kubernetes integration enabled) and `kubernetes_service`.

The target k8s cluster name is passed in the client cert, along with k8s
users/groups information.

`kubernetes_service` only serves requests for its direct k8s cluster
(from `Forwarder.creds`) and doesn't route requests to other teleport
instances.

`proxy_service` can serve requests:
- directly to a k8s cluster (the way it works pre-5.0)
- to a leaf teleport cluster (also same as pre-5.0, based on
  `RouteToCluster` field in the client cert)
- to a `kubernetes_service` (directly or over a tunnel)

The last two modes require the proxy to generate an ephemeral client TLS
cert to do an outbound mTLS connection.

# tctl flag

A flag `--kube-cluster-name` for `tctl auth sign --format=kubernetes`
which allows generating client certs for non-default k8s cluster name
(as long as it's registered in a cluster).
I used this for testing, but it could be used for automation too.

### CODE REVIEW

**Comment by awly** [comment]
This is ready for review.
I'll rebase and address @a-palchikov comments tomorrow.
There are some TODOs for unit tests, which I'll add while everyone is reviewing.

File: lib/kube/proxy/forwarder.go [hunk 1]
```diff
@@ -404,24 +421,73 @@ func (f *Forwarder) setupContext(ctx auth.Context, req *http.Request, isRemoteUs
...
+		dialFn = func(ctx context.Context, network, addr, serverID string) (net.Conn, error) {
+			return localCluster.DialTCP(reversetunnel.DialParams{
+				From:     &utils.NetAddr{AddrNetwork: "tcp", Addr: req.RemoteAddr},
+				To:       &utils.NetAddr{AddrNetwork: "tcp", Addr: addr},
+				ServerID: serverID,
+			})
+		}
+		isRemoteClosed = localCluster.IsClosed
 	} else {
-		dialFn = new(net.Dialer).DialContext
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/kube/proxy/forwarder.go:416]: the signature of `dialFn` takes an extra `serverID` argument.
`net.Dialer.DialContext` does not.
thus this has to be a wrapper here.

File: lib/reversetunnel/agentpool.go [hunk 2]
```diff
@@ -356,12 +366,29 @@ func (l ServerHandlerToListener) Close() error {
...
+	closeOnce *sync.Once
+	closed    chan struct{}
 }
 
+func newConnCloser(c net.Conn) connCloser {
+	return connCloser{Conn: c, closeOnce: new(sync.Once), closed: make(chan struct{})}
+}
+
+func (c connCloser) Close() error {
+	c.closeOnce.Do(func() { close(c.closed) })
```
Comment (a-palchikov) [inline lib/reversetunnel/agentpool.go:383]: I think this could be simpler:
```go
func (c connCloser) Close() error {
	select {
	case <-c.closed:
  	  	return nil // this way a double Close can be avoided if necessary
	default:
		close(c.closed)
  	  	return c.Conn.Close()
	}
}
```

File: lib/reversetunnel/transport.go [hunk 3]
```diff
@@ -262,23 +285,26 @@ func (p *transport) start() {
...
-			p.reply(req, false, []byte("connection rejected: server missing"))
-			return
-		}
-		if p.sconn == nil {
-			p.reply(req, false, []byte("connection rejected: server connection missing"))
-			return
-		}
+		if p.server != nil {
+			if p.sconn == nil {
+				p.log.Debug("Connection rejected: server missing")
```
Comment (a-palchikov) [inline lib/reversetunnel/transport.go:290]: message discrepancy: `server connection missing`

File: lib/auth/middleware_test.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	// Helper func for generating fake certs.
+	subject := func(id tlsca.Identity) pkix.Name {
+		s, err := id.Subject()
+		require.NoError(t, err)
+		// ExtraNames get moved to Names when generating a real x509 cert.
+		// Since we're just mimicking certs in memory, move manually.
+		s.Names = s.ExtraNames
+		return s
+	}
+	now := time.Now()
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/auth/middleware_test.go]: changed to a hardcoded date

File: lib/reversetunnel/agentpool.go (hunk 2 above)
Comment (awly) [reply lib/reversetunnel/agentpool.go:383]: I think your suggestion can still fail when multiple goroutines call `Close` at the same time.
They will both drop into the `default` clause and call `close(c.closed)` causing a panic

File: lib/reversetunnel/transport.go (hunk 3 above)
Comment (awly) [reply lib/reversetunnel/transport.go]: fixed

File: lib/reversetunnel/agentpool.go [hunk 5]
```diff
@@ -356,12 +366,29 @@ func (l ServerHandlerToListener) Close() error {
...
+func (c connCloser) wait() { <-c.closed }
+
 // reverseTunnelAddr is a net.Addr implementation for a listener based on a
 // reverse tunnel.
-type reverseTunnelAddr struct{}
+type reverseTunnelAddr string
 
-func (reverseTunnelAddr) Network() string { return "ssh" }
-func (reverseTunnelAddr) String() string  { return "ssh://kube.reversetunnel.teleport.cluster.local" }
+func (reverseTunnelAddr) Network() string  { return "ssh-reversetunnel" }
```
Comment (r0mant) [inline lib/reversetunnel/agentpool.go:393]: Not sure yet how exactly this is used in the reverse tunnel subsystem (if at all), but just making sure this won't cause backwards-incompatibility b/w old and new versions of components.

File: lib/kube/proxy/auth.go [hunk 6]
```diff
@@ -53,83 +53,107 @@ func TestOnlySkipSelfPermissionCheck(skip bool) {
...
+		log := log.WithField("cluster", cluster)
+		log.Debug("Checking kubernetes impersonation permissions.")
+		client, err := kubernetes.NewForConfig(clientCfg)
 		if err != nil {
-			return nil, trace.BadParameter(`auth server assumed that it is
-running in a kubernetes cluster, but could not init in-cluster kubernetes client: %v`, err)
+			return nil, trace.Wrap(err, "failed to generate kubernetes client for cluster %q: %v", cluster, err)
+		}
+		// For each loaded cluster, check impersonation permissions. This
+		// should only fail for newKubeService.
```
Comment (r0mant) [inline lib/kube/proxy/auth.go:114]: Nit: "This should only fail for newKubeService" gives a wrong impression, I would rephrase to "The failure is only critical error for a newKubeService".

**Review by r0mant — COMMENTED** [review]
Mostly stylistic suggestions/typos.

File: lib/auth/middleware.go [hunk 7]
```diff
@@ -412,16 +412,23 @@ func (a *Middleware) GetUser(connState tls.ConnectionState) (IdentityGetter, err
...
 	identity, err := tlsca.FromSubject(clientCert.Subject, clientCert.NotAfter)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	// Since 5.0, teleport TLS certs include the origin teleport cluster in the
+	// subject (identity). Before 5.0, origin teleport cluster was inferred
+	// from the cert issuer.
+	certClusterName := identity.TeleportCluster
+	if certClusterName == "" {
+		certClusterName, err = tlsca.ClusterName(clientCert.Issuer)
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/auth/middleware.go]: done

File: lib/kube/proxy/auth.go [hunk 8]
```diff
@@ -53,83 +53,107 @@ func TestOnlySkipSelfPermissionCheck(skip bool) {
...
-			return nil, trace.BadParameter(`auth server assumed that it is
-running in a kubernetes cluster, but %v mounted in pods could not be read: %v,
-set kubeconfig_file if auth server is running outside of the cluster`, teleport.KubeCAPath, err)
+// getKubeCreds fetches the kubernetes API credentials.
+//
+// There are 2 possible sources of credentials:
+// - pod service account credentials: files in hardcoded paths when running
+//   inside of a k8s pod; this is used when kubeClusterName is set
+// - kubeconfig: a file with a set of k8s endpoints and credentials mapped to
+//   theml this is used when kubeconfigPath is set
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/kube/proxy/auth.go]: done

File: lib/kube/proxy/auth.go [hunk 9]
```diff
@@ -53,83 +53,107 @@ func TestOnlySkipSelfPermissionCheck(skip bool) {
...
+	cfg, err := kubeutils.GetKubeConfig(kubeconfigPath, newKubeService, kubeClusterName)
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	if trace.IsNotFound(err) || len(cfg.Contexts) == 0 {
+		if newKubeService {
+			return nil, trace.BadParameter("no kubernetes credentials found; kubernetes_service requires either a valid kubeconfig_path or to run inside of a kubernetes pod")
 		}
+		log.Debugf("Could not load kubernetes credentials. This proxy will still handle kubernetes requests for trusted teleport clusters or kubernetes nodes in this teleport cluster")
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/kube/proxy/auth.go]: Changed to return an empty map instead.
There are a lot of other errors returned by this function. Something else could return a `NotFound` error and confuse the caller.

File: lib/kube/proxy/auth.go [hunk 10]
```diff
@@ -53,83 +53,107 @@ func TestOnlySkipSelfPermissionCheck(skip bool) {
...
+	res := make(map[string]*kubeCreds, len(cfg.Contexts))
+	// Convert kubeconfig contexts into kubeCreds.
+	for cluster, clientCfg := range cfg.Contexts {
+		log := log.WithField("cluster", cluster)
+		log.Debug("Checking kubernetes impersonation permissions.")
+		client, err := kubernetes.NewForConfig(clientCfg)
 		if err != nil {
-			return nil, trace.BadParameter(`auth server assumed that it is
-running in a kubernetes cluster, but could not init in-cluster kubernetes client: %v`, err)
+			return nil, trace.Wrap(err, "failed to generate kubernetes client for cluster %q: %v", cluster, err)
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/kube/proxy/auth.go]: nope, removed

File: lib/kube/proxy/auth.go (hunk 6 above)
Comment (awly) [reply lib/kube/proxy/auth.go]: done

File: lib/reversetunnel/agentpool.go (hunk 5 above)
Comment (awly) [reply lib/reversetunnel/agentpool.go:393]: this code was added in https://github.com/gravitational/teleport/pull/4611 and is not used by any components other than `kubernetes_service`.
until we release 5.0, there are no backwards-compatibility concerns for `kubernetes_service` since it's new.

**Review by awly — COMMENTED** [review]
Thanks for the reviews, PTAL

**Review by r0mant — APPROVED** [review]

File: lib/reversetunnel/agentpool.go (hunk 2 above)
Comment (a-palchikov) [reply lib/reversetunnel/agentpool.go:383]: Ok, that's true.

File: lib/kube/proxy/forwarder_test.go [hunk 11]
```diff
@@ -277,62 +332,122 @@ func (s ForwarderSuite) TestAuthenticate(c *check.C) {
...
-			c.Assert(trace.IsAccessDenied(err), check.Equals, true)
-			continue
-		} else {
-			c.Assert(err, check.IsNil)
-		}
+			gotCtx, err := f.authenticate(req)
+			if tt.wantErr {
+				require.Error(t, err)
+				require.True(t, trace.IsAccessDenied(err))
+				return
```
Comment (a-palchikov) [inline lib/kube/proxy/forwarder_test.go:440]: linter complains - `else` is not necessary following a `return`.

File: lib/auth/auth_test.go [hunk 12]
```diff
@@ -248,12 +253,104 @@ func (s *AuthSuite) TestAuthenticateSSHUser(c *C) {
...
+		Expires:           gotTLSCert.NotAfter,
+		RouteToCluster:    "me.localhost",
+		TeleportCluster:   "me.localhost",
+	}
+	gotID, err = tlsca.FromSubject(gotTLSCert.Subject, gotTLSCert.NotAfter)
+	c.Assert(err, IsNil)
+	c.Assert(*gotID, DeepEquals, wantID)
+
+	// Login specifying an invalid kube cluster. This should fail.
+	resp, err = s.a.AuthenticateSSHUser(AuthenticateSSHRequest{
```
Comment (a-palchikov) [inline lib/auth/auth_test.go:343]: linter: unused `resp`

**Review by a-palchikov — APPROVED** [review]
lgtm but take it with a grain of salt as I could've missed an architectural issue.

File: lib/auth/apiserver.go [hunk 13]
```diff
@@ -140,8 +140,11 @@ func NewAPIServer(config *APIConfig) http.Handler {
 	srv.DELETE("/:version/tunnelconnections/:cluster/:conn", srv.withAuth(srv.deleteTunnelConnection))
 	srv.DELETE("/:version/tunnelconnections/:cluster", srv.withAuth(srv.deleteTunnelConnections))
 	srv.DELETE("/:version/tunnelconnections", srv.withAuth(srv.deleteAllTunnelConnections))
+	// TODO(awly): migrate these to the gRPC service.
 	srv.POST("/:version/kube_services", srv.withAuth(srv.upsertKubeService))
 	srv.GET("/:version/kube_services", srv.withAuth(srv.getKubeServices))
+	srv.DELETE("/:version/kube_services/:name", srv.withAuth(srv.deleteKubeService))
+	srv.DELETE("/:version/kube_services", srv.withAuth(srv.deleteAllKubeServices))
```
Comment (fspmarshall) [inline lib/auth/apiserver.go:147]: If these are new, why not start them off as gRCP methods?

**Review by fspmarshall — APPROVED** [review]

File: lib/auth/apiserver.go (hunk 13 above)
Comment (awly) [reply lib/auth/apiserver.go:147]: I wrote these before realizing my mistake :)
Rewrite to gRPC should happen before the release anyway, I'd rather get this PR in as is and make a follow-up
